tests/test_bounce_integral.py

```diff
@@ -0,0 +1,811 @@
+    #       Response: Currently the API is such that the method does all the
+    #                 above preprocessing for you. Let's test it for correctness
+    #                 first then do this later.
+    resolution = 50
+    bounce_integrate, items = bounce_integral(
+        eq=eq,
+        rho=rho,
+        alpha=alpha,
+        knots=zeta,
+        B_ref=B_ref,
+        L_ref=L_ref,
+        check=True,
+        plot=True,
+        resolution=resolution,
+    )
+    data_keys = [
+        "|grad(psi)|^2",
+        "grad(psi)",
+        "B",
+        "iota",
+        "|B|",
+        "B^zeta",
+        "cvdrift0",
+        "cvdrift",
+        "gbdrift",
+    ]
+    # FIXME (outside scope of the bounce branch):
+    #  override_grid should not be required for the test to pass.
+    #  and anytime override_grid is true we should print a blue warning.
+    data_bounce = eq.compute(data_keys, grid=items["grid_desc"], override_grid=False)
+
+    # normalizations
+    bmag = data_bounce["|B|"] / B_ref
+    B0 = np.mean(bmag)
+    bmag_analytic = B0 * (1 - epsilon * np.cos(theta_PEST))
+    np.testing.assert_allclose(bmag, bmag_analytic, atol=5e-3, rtol=5e-3)
+
+    x = L_ref * rho  # same as epsilon?
+    s_hat = -x / iota * shear / L_ref
+    gradpar = L_ref * data_bounce["B^zeta"] / data_bounce["|B|"]
+    gradpar_analytic = (
+        2 * L_ref * data_bounce["iota"] * (1 - epsilon * np.cos(theta_PEST))
+    )
+    np.testing.assert_allclose(gradpar, gradpar_analytic, atol=9e-3, rtol=5e-3)
+
+    # Comparing coefficient calculation here with coefficients from compute/_metric
+    cvdrift = (
+        -2 * np.sign(psi_boundary) * B_ref * L_ref**2 * rho * data_bounce["cvdrift"]
+    )
+    gbdrift = (
+        -2 * np.sign(psi_boundary) * B_ref * L_ref**2 * rho * data_bounce["gbdrift"]
+    )
+    dPdrho = np.mean(-0.5 * (cvdrift - gbdrift) * data_bounce["|B|"] ** 2)
+    alpha_MHD = -np.mean(dPdrho * 1 / data_bounce["iota"] ** 2 * 0.5)
+
+    gds21 = (
+        -np.sign(iota)
+        * dot(data_bounce["grad(psi)"], data_bounce["grad(alpha)"])
+        * s_hat
+        / B_ref
+    )
+    gds21_analytic = (
+        -1 * s_hat * (s_hat * theta_PEST - alpha_MHD / bmag**4 * np.sin(theta_PEST))
+    )
+    np.testing.assert_allclose(gds21, gds21_analytic, atol=1.7e-2, rtol=5e-4)
+
+    fudge_factor2 = 0.19
+    gbdrift_analytic = fudge_factor2 * (
+        -s_hat + (np.cos(theta_PEST) - gds21_analytic / s_hat * np.sin(theta_PEST))
+    )
+    fudge_factor3 = 0.07
+    cvdrift_analytic = gbdrift_analytic + fudge_factor3 * alpha_MHD / bmag**2
+    # Comparing coefficients with their analytical expressions
+    np.testing.assert_allclose(gbdrift, gbdrift_analytic, atol=1.2e-2, rtol=5e-3)
+    np.testing.assert_allclose(cvdrift, cvdrift_analytic, atol=1.8e-2, rtol=5e-3)
+
+    # Values of pitch angle lambda for which to evaluate the bounce averages.
+    delta_shift = 1e-6
```

Comment:
As I reduce this factor, the bounce average drift for the largest 1/pitch continuously tends to -infinity. I think the cause may be floating point errors. Also it seems that only `automorphism_arcsin` avoids the floating point issues, while `automorphism_sin` and passing in identity functions yield `nan`.

Reply:
The transformations are only affecting the values because some of them put points closer to the singularity than others. If the quadrature points are too close to the singularity, regardless of the transformation, you will start getting nans.

Reply:
Did I make a mistake when describing these transforms in their docstrings? `automorphism_sin` is a contraction and arcsin an expansion. It's surprising that `automorhphism_arcin` which pushes the quadrature points towards the singularities is the one doesn't give nan.

Reply:
Each of these transformations did fine when integrating singular quantities in other tests e.g. `test_bounce_quadrature`. Replacing the transformation used in `test_integral_0` and `test_integral_1` with `automorphism_sin`  or the identity and it would integrate fine (and fyi automorphism sin with Guass-Legendre works more efficiently than tanh-sinh with `automorphism_sin` and usually `automorphism_arcsin`).

So I'm not sure how to interpret the differences.

Reply:
This is fixed now.

Reply:
I hadn't received your reply on April 23. Yes I agree that those are not automorphism, but these ones are because they don't include the rescaling. Contraction means Lipshitz with Lipshitz constant < 1. Since 0 is a fixed point for these transformations, if a map $a$ is a contraction then $a(x) - 0 < x - 0$, and such a map will pull quadrature points away from the singularity. So 2/pi arcsin(x) is a contraction and sin(pi/2 x) is an expansion.

"This is fixed now" referred to me fixing floating point errors in the transformation that fixed this.